1-Draft/RFCNNNN-Supports-Parallel.md

```diff
@@ -0,0 +1,316 @@
+qualifier was used on them.
+* update the command processor to support invocation of the `ProcessRecord`
+method inside of a `ThreadJob` when the command is invoked with one of the new
+common parameters.
+* update the command processor such that it waits for `ThreadJob` instances it
+creates to complete, and then writes their results to the pipeline as they
+finish execution.
+* update `ForEach-Object` to be the first command that support `SupportsParallel`,
+so that users have an easy way to create multi-threaded, multi-runspace
+execution paths in PowerShell.
+
+## Alternate Proposals and Considerations
+
+### Running a parallelized command asynchronously
+
+Parallelization is about being able to perform multi-threaded processing of
+objects in a pipeline. This is what is natively supported by the enhancements
+outlined in this RFC. The RFC that originally inspired this approach included
+an `-AsJob` parameter that would allow for asynchronous invocation of the process
+block of `ForEach-Object`. There are two problem with `-AsJob` in this scenario:
+
+1. (Minor) Other commands use `-AsJob` as a parameter today to return a
+background job that is run in a separate child process, which in turn may
+contain multiple child background jobs. Having this parameter also used to
+return one or more thread jobs is bound to confuse some people. The solution
+here is straightforward: use a `-AsThreadJob` parameter name instead of `-AsJob`.
+1. (Major) Pipelines in PowerShell have a very specific processing sequence. In
+general, first the `begin` blocks in the pipeline are processed, then the `process`
+blocks, and last the `end` blocks. In the future we will also see a `dispose` block
+that gets processed after the end block, or on terminating error. Introducing a
+parameter that makes a command optionally return a job from a process block means
+that command cannot properly handle logic in the `end` or `dispose` blocks,
+because those should only be invoked after the `process` block work has
+finished execution, and returning a job defers that execution until later.
+
+The second issue is so significant that it highlights a very important detail
+when considering adding parallel support to a command:
+
+_Commands that support parallel processing must be run to completion so that
+`end` and `dispose` blocks do their work in the proper sequence, and any
+decision to make the pipeline in which those commands are used run as a
+background job or as a thread job needs to be made outside of the pipeline._
+
+With that in mind, parameters related to creating a job have been excluded from
+this RFC, and a related RFC has been created to facilitate invocation of a
+pipeline in a thread job by using an operator that is external to the pipeline.
+
+### Runspace pooling (and other possible future parallel improvements)
+
+It may be desired to allow users to create a pool of runspaces, and then cycle
```

Comment:
>It may be desired to allow users to create a pool of runspaces...

I think this is an understatement.  From the testing I have performed any parallelization without sophisticated runspace initialization and reuse is basically unusable save for a few contrived scenarios.  Without such runspace management, for most scenarios so much time is spent waiting for modules to load that that wait far surpasses the gains of parallelization.  I think [this repro from PowerShell/PowerShell#7524](https://github.com/PowerShell/PowerShell/issues/7524#issuecomment-414520822) is the most direct demonstration of the problem.